// @appsignal/nodejs/dist/logger.js
"use strict";
var _client, _group;
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseLogger = exports.LOGGER_LEVEL_SEVERITY = void 0;
const tslib_1 = require("tslib");
exports.LOGGER_LEVEL_SEVERITY = {
    trace: 1,
    debug: 2,
    info: 3,
    log: 4,
    warn: 5,
    error: 6
};
const UNKNOWN_SEVERITY = 0;
function severity(level) {
    var _a;
    return (_a = exports.LOGGER_LEVEL_SEVERITY[level]) !== null && _a !== void 0 ? _a : UNKNOWN_SEVERITY;
}
class BaseLogger {
    constructor(client, group, level = "info") {
        _client.set(this, void 0);
        _group.set(this, void 0);
        if (typeof group != "string") {
            throw new TypeError(`Logger group name must be a string; ${typeof group} given`);
        }
        tslib_1.__classPrivateFieldSet(this, _client, client);
        tslib_1.__classPrivateFieldSet(this, _group, group);
        this.severityThreshold = severity(level);
        if (this.severityThreshold == UNKNOWN_SEVERITY) {
            tslib_1.__classPrivateFieldGet(this, _client).integrationLogger.warn(`Logger level must be "trace", "debug", "info", "log", "warn" or "error", ` +
                `but "${level}" was given. Logger level set to "info".`);
            this.severityThreshold = severity("info");
        }
    }
    trace(message, attributes) {
        this.sendLog(severity("trace"), message, attributes);
    }
    debug(message, attributes) {
        this.sendLog(severity("debug"), message, attributes);
    }
    info(message, attributes) {
        this.sendLog(severity("info"), message, attributes);
    }
    log(message, attributes) {
        this.sendLog(severity("log"), message, attributes);
    }
    warn(message, attributes) {
        this.sendLog(severity("warn"), message, attributes);
    }
    error(message, attributes) {
        this.sendLog(severity("error"), message, attributes);
    }
    sendLog(severity, message, attributes = {}) {
        if (severity < this.severityThreshold) {
            return;
        }
        tslib_1.__classPrivateFieldGet(this, _client).extension.log(tslib_1.__classPrivateFieldGet(this, _group), severity, String(message), attributes);
    }
}
exports.BaseLogger = BaseLogger;
_client = new WeakMap(), _group = new WeakMap();
